fendermint/app/src/app.rs

```diff
@@ -870,13 +877,47 @@ where
 
         if let Some(ref client) = self.snapshots {
             match atomically_or_err(|| {
-                client.apply_chunk(request.index, request.chunk.clone().into())
+                client.save_chunk(request.index, request.chunk.clone().into())
             })
             .await
             {
-                Ok(completed) => {
-                    if completed {
-                        tracing::info!("received all snapshot chunks");
+                Ok(snapshot) => {
+                    if let Some(snapshot) = snapshot {
+                        tracing::info!(
+                            download_dir = snapshot.snapshot_dir.to_string_lossy().to_string(),
+                            height = snapshot.manifest.block_height,
+                            "received all snapshot chunks",
+                        );
+
+                        // Ideally we would import into some isolated store then validate,
+                        // but for now let's trust that all is well.
+                        if let Err(e) = snapshot.import(self.state_store_clone(), true).await {
+                            tracing::error!(error =? e, "failed to import snapshot");
+                            return Ok(response::ApplySnapshotChunk {
+                                result: response::ApplySnapshotChunkResult::RejectSnapshot,
+                                ..default
+                            });
+                        }
+
+                        tracing::info!(
+                            height = snapshot.manifest.block_height,
+                            "imported snapshot"
+                        );
+
+                        // Now insert the new state into the history.
+                        let mut state = self.committed_state()?;
+                        state.block_height = snapshot.manifest.block_height;
+                        state.state_params = snapshot.manifest.state_params;
+                        self.set_committed_state(state)?;
+
+                        // TODO: We can remove the `current_download` from the STM
+                        // state here which would cause it to get dropped from /tmp,
+                        // but for now let's keep it just in case we need to investigate
+                        // some problem.
+
+                        // We could also move the files into our own snapshot directory
```

Comment:
To be honest it's mostly laziness, but I also think we should first try to resume syncing with the chain before we re-transmit this snapshot. I'm not exactly sure whether it's possible to do too much validation on the snapshot contents. I validate the checksum, but I am not sure whether a particular snapshot chosen by CometBFT necessarily has more than one validator behind it, or it might be just a one-off, which would allow somebody to smuggle in illegal CIDs into the state store, and then later cause consensus failure by deploying a contract that looks it up, finding it one the store of some, but not other validators. 

IIRC this would only affect Wasm actors, not EVM, and by the time it's relevant the FVM might offer protection against it. But it reminds me that I should try to remove even the metadata record from the blockstore after the import.